### PR TITLE
#164797111 ft edit user profile

### PR DESCRIPTION
**Pull request by emmsdan**
#### What does this PR do?
- allows the user to update their profile information

#### Description of Task to be completed?
none

#### How should this be manually tested?
- clone this repo and checkout to `ft-edit-user-profile-164797111`
- run `npm i & npm run watch`
- navigate to `http://locahost:8080/users`
- edit a profile

#### Any background context you want to provide?
none

#### What are the relevant pivotal tracker stories? (if applicable)
#164797111

### CODE REVIEW

File: app.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (Japhethca) [inline app.json]: Since this file is not part of this PR, we can remove it.

File: package.json [hunk 2]
```diff
@@ -23,7 +23,7 @@
     ]
   },
   "scripts": {
-    "test": "jest",
+    "test": "jest --coverage",
```
Comment (Japhethca) [inline package.json]: We can add a new script to handle test coverage, like `test:coverage`.

File: package.json [hunk 3]
```diff
@@ -40,8 +40,17 @@
       "./src/setupTests.js"
     ],
     "moduleNameMapper": {
-      "^.+\\.(css|less|scss)$": "babel-jest"
-    }
+      "^.+\\.(css|less|scss)$": "babel-jest",
+      "^.+\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "babel-jest"
```
Comment (Japhethca) [inline package.json]: we can remove the file extensions that we are not using.

File: package.json [hunk 4]
```diff
@@ -97,14 +106,18 @@
...
     "react-redux": "^7.0.3",
     "react-router": "^5.0.0",
     "react-router-dom": "^5.0.0",
     "redux": "^4.0.1",
+    "redux-mock-store": "^1.5.3",
+    "redux-thunk": "^2.3.0",
     "semantic-ui-css": "^2.4.1",
-    "semantic-ui-react": "^0.86.0"
+    "semantic-ui-react": "^0.86.0",
+    "toastr": "^2.1.4"
```
Comment (Japhethca) [inline package.json]: We can update this to `react-toastify`.

File: src/components/containers/ArticleLists/ArticleLists.jsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import React, { Component } from 'react';
+import { Item, Icon } from 'semantic-ui-react';
+
+import Title from '../../presentationals/Title/Title.jsx';
+
+const f16 = {
+  fontSize: '16px',
+};
+
+class ArticleLists extends Component {
```
Comment (Japhethca) [inline src/components/containers/ArticleLists/ArticleLists.jsx]: We can change this to a stateless component.

File: src/components/containers/ArticleLists/ArticleLists.jsx [hunk 6]
```diff
@@ -0,0 +1,95 @@
+/* eslint-disable react/jsx-key */
```
Comment (Japhethca) [inline src/components/containers/ArticleLists/ArticleLists.jsx]: let's remove this file since its not used used here.

File: src/components/containers/ProfileMain/ProfileMain.jsx [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+/* eslint-disable react/display-name */
+import React, { Component } from 'react';
+import { Tab, Icon } from 'semantic-ui-react';
+
+import Title from '../../presentationals/Title/Title.jsx';
+import ArticleLists from '../ArticleLists/ArticleLists.jsx';
+import Settings from '../ProfileSettings/ProfileSettings.jsx';
+import './ProfileMain.scss';
+
+class ProfileMain extends Component {
```
Comment (Japhethca) [inline src/components/containers/ProfileMain/ProfileMain.jsx]: We can change this to stateless component.

File: src/components/containers/ProfileMain/ProfileMain.jsx [hunk 8]
```diff
@@ -0,0 +1,126 @@
+/* eslint-disable react/prop-types */
+/* eslint-disable react/display-name */
+import React, { Component } from 'react';
+import { Tab, Icon } from 'semantic-ui-react';
+
+import Title from '../../presentationals/Title/Title.jsx';
+import ArticleLists from '../ArticleLists/ArticleLists.jsx';
+import Settings from '../ProfileSettings/ProfileSettings.jsx';
```
Comment (Japhethca) [inline src/components/containers/ProfileMain/ProfileMain.jsx]: Lets rename this to `ProfileSettings`.

File: src/components/containers/ProfileSettings/ProfileSettings.jsx [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+import Button from '../../presentationals/Button/Button.jsx';
+import { generateFormData, toastbar } from '../../../utils/helpers';
+import {
+  updateUserProfileRequest,
+  uploadImageToServer,
+} from '../../../store/modules/Profile';
+import Loader from '../../presentationals/Loader/Loader.jsx';
+
+class Settings extends Component {
```
Comment (Japhethca) [inline src/components/containers/ProfileSettings/ProfileSettings.jsx]: Let's rename this component to `ProfileSetting`

File: src/components/containers/ProfileSettings/ProfileSettings.jsx [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+    reader.readAsDataURL(image);
+    toastbar.success('Photo Ready to be saved to profile', 'DONE');
+    this.setState({ previewImage: image });
+  };
+
+  render() {
+    return (
+      <div className="settings">
+        <div className="edit-profile">
+          {this.state.isLoading ? (
```
Comment (Japhethca) [inline src/components/containers/ProfileSettings/ProfileSettings.jsx]:
```javascript 

{this.state.isLoading} && <Loader />

File: src/components/containers/ProfileSettings/ProfileSettings.jsx [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+  }
+}
+
+const editButtonValues = {
+  type: 'submit',
+  value: 'SAVE CHANGES',
+  class: 'btn-dark',
+};
+
+const mapStateToProps = state => {
```
Comment (Japhethca) [inline src/components/containers/ProfileSettings/ProfileSettings.jsx]: this can be rewritten as
```javascript
const mapStateToProps = state => ({profile: state.profileReducer.profile});

File: src/components/containers/ProfileSidebar/ProfileSidebar.jsx [hunk 12]
```diff
@@ -0,0 +1,74 @@
+/* eslint-disable no-console */
```
Comment (Japhethca) [inline src/components/containers/ProfileSidebar/ProfileSidebar.jsx]: lets remove these eslint comments.

File: src/components/containers/ProfileSidebar/ProfileSidebar.jsx [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+/* eslint-disable react/prop-types */
+import React, { Component } from 'react';
+import { Icon } from 'semantic-ui-react';
+
+import Avatar from '../../presentationals/Avatar/Avatar.jsx';
+import Button from '../../presentationals/Button/Button.jsx';
+
+import './ProfileSidebar.scss';
+
+class ProfileSidebar extends Component {
```
Comment (Japhethca) [inline src/components/containers/ProfileSidebar/ProfileSidebar.jsx]: This should be a functional component.

File: src/components/containers/ProfileSidebar/ProfileSidebar.jsx [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+  constructor(props) {
+    super(props);
+  }
+  render() {
+    let userAvatar = {
+      url:
+        this.props.profile.image ||
+        'https://res.cloudinary.com/nedy123/image/upload/v1557149386/avatar_pupqrh.svg',
```
Comment (Japhethca) [inline src/components/containers/ProfileSidebar/ProfileSidebar.jsx]: let put this image url in our config file.

File: src/components/containers/ProfileSidebar/ProfileSidebar.jsx [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+      <div className="overlay-image">
+        <CameraIcon />
+      </div>
+    );
+
+    const { followers, followings, username, email } = this.props.profile;
+    return (
+      <div className="sidebar">
+        <div className="avatar-div">
+          <Avatar avatar={userAvatar} />
```
Comment (Japhethca) [inline src/components/containers/ProfileSidebar/ProfileSidebar.jsx]: We can rewrite the Avatar` component to use its props directly as shown below
```javascript 
<Avatar 
    url={} 
    className={} 
/>

File: src/pages/Profile.jsx [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+import { connect } from 'react-redux';
+
+import Navbar from '../components/presentationals/Navbar/Navbar.jsx';
+import ProfileSidebar from '../components/containers/ProfileSidebar/ProfileSidebar.jsx';
+import ProfileMain from '../components/containers/ProfileMain/ProfileMain.jsx';
+import './Profile.scss';
+import { getUserProfileRequest } from '../store/modules/Profile';
+
+class Profile extends Component {
+  constructor(props) {
```
Comment (Japhethca) [inline src/pages/Profile.jsx]: the constructor is redundant here.

File: src/store/modules/Profile.js [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+  UPDATE_IMAGE_ERROR,
+  UPDATE_IMAGE_SUCCESS,
+} from '../actions/types';
+import { setToken, http } from '../../utils/helpers';
+import { getToken } from '../../utils/helpers';
+import updateProfileRequest from '../actions/updateProfileActions';
+import { toastbar } from '../../utils/helpers';
+
+// for development use only
+setToken(
```
Comment (Japhethca) [inline src/store/modules/Profile.js]: lets remove this set token function call

File: app.json (hunk 1 above)
Comment (emmsdan) [reply app.json]: Ok, I will do just that. thanks

File: package.json (hunk 2 above)
Comment (emmsdan) [reply package.json]: Ok.

File: package.json (hunk 4 above)
Comment (emmsdan) [reply package.json]: Ok.

File: src/components/containers/ArticleLists/ArticleLists.jsx (hunk 5 above)
Comment (emmsdan) [reply src/components/containers/ArticleLists/ArticleLists.jsx]: Ok, removed.

File: src/components/containers/ProfileMain/ProfileMain.jsx (hunk 7 above)
Comment (emmsdan) [reply src/components/containers/ProfileMain/ProfileMain.jsx]: I have changed this.

File: src/components/containers/ProfileMain/ProfileMain.jsx (hunk 8 above)
Comment (emmsdan) [reply src/components/containers/ProfileMain/ProfileMain.jsx]: Ok. I have renamed it.

File: src/components/containers/ProfileSettings/ProfileSettings.jsx (hunk 9 above)
Comment (emmsdan) [reply src/components/containers/ProfileSettings/ProfileSettings.jsx]: I have renamed it.

File: src/components/containers/ProfileSettings/ProfileSettings.jsx (hunk 10 above)
Comment (emmsdan) [reply src/components/containers/ProfileSettings/ProfileSettings.jsx]: ok.

File: src/components/containers/ProfileSidebar/ProfileSidebar.jsx (hunk 14 above)
Comment (emmsdan) [reply src/components/containers/ProfileSidebar/ProfileSidebar.jsx]: I have moved the URL into the config.js file

File: src/pages/Profile.jsx (hunk 16 above)
Comment (emmsdan) [reply src/pages/Profile.jsx]: ok.

File: src/components/containers/ProfileSidebar/ProfileSidebar.jsx (hunk 13 above)
Comment (emmsdan) [reply src/components/containers/ProfileSidebar/ProfileSidebar.jsx]: I have changed this.

File: src/components/containers/ProfileSidebar/ProfileSidebar.jsx (hunk 15 above)
Comment (emmsdan) [reply src/components/containers/ProfileSidebar/ProfileSidebar.jsx]: I have rewritten that.

File: src/components/containers/ProfileSidebar/ProfileSidebar.jsx (hunk 12 above)
Comment (emmsdan) [reply src/components/containers/ProfileSidebar/ProfileSidebar.jsx]: removed.

File: src/components/containers/ArticleLists/ArticleLists.jsx (hunk 6 above)
Comment (emmsdan) [reply src/components/containers/ArticleLists/ArticleLists.jsx]: done, thanks.

File: src/components/presentationals/ProfileSettings/ProfileSettings.jsx [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+  lastName: PropTypes.string,
+  username: PropTypes.string,
+  email: PropTypes.string,
+  updatedAt: PropTypes.string,
+  bio: PropTypes.string,
+};
+
+const mapStateToProps = state => {
+  const { data } = state.profile.profile;
+  return data;
```
Comment (igbominadeveloper) [inline src/components/presentationals/ProfileSettings/ProfileSettings.jsx]: I think you have to return the actual data you want to map your state to from the parent provided props for the props validation to work well

```javascript
mapStateToProps = state => ({
profile: {
  currentView: state.profile.profile.currentView,
  uploadImageToServer: state.profile.profile.,
  updateUserProfileRequest: state.profile.profile.updateUserProfileRequest,
  firstName: state.profile.profile.firstName,
  lastName: state.profile.profile.lastName,
  username: state.profile.profile.username,
  email: state.profile.profile.email,
  updatedAt: state.profile.profile.updatedAt,
  bio: state.profile.profile.bio,
}
});
```

OR 
```javascript
mapStateToProps = state => ({
profile: state.profile.profile});
```
Then, you can destructure the specific ones you want just before the return statement inside your return statement

Comment (emmsdan) [reply src/components/presentationals/ProfileSettings/ProfileSettings.jsx]: yeah,  thanks.
I will try that.

File: src/store/actions/updateProfileActions.js [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+const UPDATE_USER_PROFILE_ERROR = 'UPDATE_USER_PROFILE_ERROR';
+const UPDATE_USER_PROFILE_SUCCESS = 'UPDATE_USER_PROFILE_SUCCESS';
+
+export default function updateProfileRequest(
+  type = null,
+  response = null,
+  constantType = null,
+) {
+  switch (type) {
+    case null:
```
Comment (Japhethca) [inline src/store/actions/updateProfileActions.js]: lets use `default` case instead of checking for `null`.

File: src/utils/helpers.js [hunk 20]
```diff
@@ -1,5 +1,58 @@
...
+export function getToken() {
+  const token = localStorage.getItem('token');
+  return JSON.parse(token);
+}
+
+export const destroyToken = () => {
+  return localStorage.removeItem('token');
+};
+
+export const getUserId = () => {
```
Comment (Japhethca) [inline src/utils/helpers.js]: let's remove this function since we are not using it.

File: src/components/presentationals/ProfileMain/ProfileMain.test.jsx [hunk 21]
```diff
@@ -82,7 +82,7 @@ describe('ProfileMain', () => {
 
   it('should render Settings Tab correctly', () => {
     const component = mount(<ProfileMain {...props} />);
-    const login = component.find('MenuItem[name="Settings"]').simulate('click');
-    expect(login).toMatchSnapshot();
+    // const login = component.find('MenuItem[name="Settings"]').simulate('click');
```
Comment (Japhethca) [inline src/components/presentationals/ProfileMain/ProfileMain.test.jsx]: Remove commented code.

Comment (emmsdan) [reply src/components/presentationals/ProfileMain/ProfileMain.test.jsx]: Nice catch, I have removed it.

**Review by Japhethca — APPROVED** [review]
